MSC2078: Sending Third-Party Request Tokens via the Homeserver (#2078)
matrix-org/matrix-doc: proposals/2078-homeserver-password-resets.md

# MSC2078 - Sending Third-Party Request Tokens via the Homeserver

This MSC proposes removing the current requirement of the identity server to
send third-party request tokens, and allows homeservers to implement the
functionality instead. These request tokens are used to verify the identity of
the request auther as an owner of the third-party identity (3PID). This can be

> used for binding a 3PID to an account, or for resetting passwords via email or

Reviewer: Erm: if it is used for binding a 3pid to an account, is it still done through  the `/_matrix/client/r0/account/password/email/requestToken` API? If so, it might be worth calling out that fact in the doc because it confused the hell out of me.

Author: Oh, right. That is done via the `/_matrix/client/r0/account/3pid/email/requestToken` endpoint. I'm going to need to refactor the doc a bit to mention that it should be for requesting any token,

The intention is to put less trust in the identity server, which is currently
one of the most centralised components of Matrix. As it stands, an attacker in
control of a identity server can reset a user's password if the identity server
is considered trusted by that homeserver, and the user has registered at least
one 3PID. This is due to the identity server currently handling the job of
confirming the user's control of that identity.

The MSC aims to simply clarify that homeservers can take on the responsibility
of sending password reset tokens themselves.

## Proposal

Currently when a client requests a password reset, it makes a call to either
[/_matrix/client/r0/account/password/email/requestToken](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-account-password-email-requesttoken)
or
[/_matrix/client/r0/account/password/msisdn/requestToken](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-account-password-msisdn-requesttoken).
This request is supplied all the necessary details as well as an `id_server`
field containing the address of a identity server trusted by the homeserver.

> The `id_server` field is currently required as the homeserver must know where

Reviewer: for future ref: this whole paragraph is kinda hard to grok.

Suggest losing the first two sentences, and rather than starting with what the MSC is *not* doing, say what it *is* doing - *then* emphasise that the `id_server` field will still be mandatory in a second paragraph.

to proxy the request to. This MSC proposes not to change the requirements of
this field. Instead, it asks to clarify that the homeserver is allowed to not
proxy the request, but carry it out itself. This would mean the homeserver can

> both send password reset tokens (via email or SMS), as well as accept requests

Reviewer: "as well as accept requests an endpoint" ?

Author: Missed an "on".

an endpoint (with the same parameters as
[/_matrix/identity/api/v1/validate/email/submitToken](https://matrix.org/docs/spec/identity_service/r0.1.0.html#post-matrix-identity-api-v1-validate-email-submittoken))
to verify that token.

> Consideration was taken not to make `id_server` and optional field. Let's

Reviewer: s/and/an/

> Consideration was taken not to make `id_server` and optional field. Let's

Reviewer: I wrote a bit of a grumble about how we ought to plan to get rid of this, and only discovered the Future Considerations section at the very end. I'd be inclined to move those future considerations in here - especially if this moves to a 'dismissed alternatives' section.

`/requestToken` omitting the `id_server` field. The homeserver however has
opted to continue proxying `/requestToken` to the identity server, even though
it knows this is potentially insecure. The homeserver now has no idea which
identity server to proxy the request to, and must return a failure to the
client. The client could then make another request with an `id_server`, but
we've now made two requests that ended up in the same outcome, instead of one,
in hopes of saving a very small amount of bandwidth by omitting the field
originally.

An additional complication is that in the case of SMS, a full link to reset
passwords is not sent, but a short code. The client then asks the user to enter
this code, however the client may now not know where to send the code. Should
it send it to the identity server or the homeserver? Which sent out the code?

In order to combat this problem, the field `submit_url` should be added in the
response from both the email and msisdn variants of the `/requestToken`
Client-Server API, if and only if the verification message contains a code the
user is expected to enter into the client (for instance in the case of a short

> code through SMS). It SHOULD be in the form of

Reviewer: s/It SHOULD be/It is suggested that it be/ ?

Author: Yeah, in Synapse we've made it `/_synapse/password_resets/email/submitToken`, so...

> `/_matrix/identity/api/v1/validate/{3pid_type}/submitToken`, similar to the

Reviewer: so this needs changing, as per https://github.com/matrix-org/matrix-doc/pull/2078#issuecomment-499444845?

server which is currently one of the most centralised components of Matrix. As
it stands, an attacker in control of a identity server can reset a user's
password if the identity server is considered trusted by that homeserver, and
the user has registered at least one third-party identifier (3PID). This is due
to the identity server currently handling the job of confirming the user's
control of that identity.

The MSC aims to simply clarify that homeservers can take on the responisibility
of sending password reset tokens themselves.

## Proposal

Currently when a client requests a password reset, they make a call to either
[/_matrix/client/r0/account/password/email/requestToken](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-account-password-email-requesttoken)
or
[/_matrix/client/r0/account/password/msisdn/requestToken](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-account-password-msisdn-requesttoken).
This request is supplied all the necessary details as well as a `id_server`
field containing the address of a trusted identity server which the user has

> used in the past to bind their 3PID.

Reviewer 2: any trusted ID server - they could have used a different one to associate it

(also 'bind' == the 3pid -> mxid linking on the IS: the mxid -> 3pid links kept by the HSs are... some less well defined term - linked 3pids or associated 3pids).

The `id_server` field is currently required as the homeserver must know where
to proxy the request to. This MSC proposes not to change the requirements of
this field. Instead, it asks to clarify that the homeserver is allowed to not
proxy the request, but carry it out itself. This would mean the homeserver can
both send password reset tokens (via email or SMS), as well as accept requests
to

> [/_matrix/identity/api/v1/validate/email/submitToken](https://matrix.org/docs/spec/identity_service/r0.1.0.html#post-matrix-identity-api-v1-validate-email-submittoken)

Reviewer 2: Probably just a URL of the HS's choosing (rather than requiring that it accept requests on a endpoint in the IS namespace).

to verify that token.

> An additional complication is that in the case of SMS, a full link to reset passwords is not sent, but a short code. The client then asks the user to enter this code, however the client may now not know where to send the code. Should it send it to the identity server or the homeserver? Which sent out the code?

Reviewer 2: line wrapping

> In order to combat this problem, the field `submit_url` should be added in the response from both the email and msisdn variants of the `/requestToken` Client-Server API, if and only if the verification message contains a code the user is expected to enter into the client (for instance in the case of a short code through SMS). If this field is omitted, the client should continue the same behaviour from before, which is to send the token to the identity server directly. This is intended for backwards compatibility with older servers.

Reviewer 2: If it's explicitly for backwards-compat it implies that if the server is proxying, is is supposed to make a `submit_url` by appending the `.../submitToken` to the id_server param - might be good to be explicit.

Author: In the future the homeserver may send `submit_url` with the homeserver's URL in front instead of the IS.

Reviewer 2: but if I'm writing an HS and want to just continue proxying the requests, do I send a `submit_url` pointing to the IS or should I just continue not sending `submit_url`?

Author: I was just saying that the HS can also send a `submit_url` with the HS's domain in front, as the HS will handle msisdn in the future.

It doesn't matter if you're proxying the request or not, you should send a `submit_url` if the client requires sending a `POST` request to complete verification.

Reviewer 2: OK - so that sounds like, "all HSes should now send `submit_url`" so was basically just suggesting spelling out that if you proxy the request, you should send the IS's URL.

Author: Right, but it's "all HSes should now send `submit_url` if they require the client to make the submission". Just want to make clear we're both on the same page.

Can definitely make that last part explicit though.

Reviewer 2: Yeah, I guess it does say they should - might still be nice to be explicit though.

Author: Yes the content of `id_server` should be ignored if you get `submit_url`. Have updated to be more explicit.

# MSC2078 - Sending Password Reset Emails via the Homeserver

This MSC proposes removing the current requirement of the identity server to
send password reset tokens, and allows homeservers to implement the
functionality instead. The intention is to put less trust in the identity
server which is currently one of the most centralised components of Matrix. As
it stands, an attacker in control of a identity server can reset a user's
password if the identity server is considered trusted by that homeserver, and
the user has registered at least one third-party identifier (3PID). This is due
to the identity server currently handling the job of confirming the user's
control of that identity.

The MSC aims to simply clarify that homeservers can take on the responisibility
of sending password reset tokens themselves.

## Proposal

> Currently when a client requests a password reset, they make a call to either

Reviewer 3: nitpick
```suggestion
Currently when a client requests a password reset, it makes a call to either
```

[/_matrix/client/r0/account/password/email/requestToken](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-account-password-email-requesttoken)
or
[/_matrix/client/r0/account/password/msisdn/requestToken](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-account-password-msisdn-requesttoken).
This request is supplied all the necessary details as well as an `id_server`
field containing the address of a identity server trusted by the homeserver.

The `id_server` field is currently required as the homeserver must know where

> to proxy the request to. This MSC proposes not to change the requirements of

Reviewer 4: should put the explanation of "making the field optional would require the homeserver return an error, which the client may not understand" somewhere in here as rationale for keeping it "required".

this field. Instead, it asks to clarify that the homeserver is allowed to not
proxy the request, but carry it out itself. This would mean the homeserver can
both send password reset tokens (via email or SMS), as well as accept requests
an endpoint (with the same parameters as
[/_matrix/identity/api/v1/validate/email/submitToken](https://matrix.org/docs/spec/identity_service/r0.1.0.html#post-matrix-identity-api-v1-validate-email-submittoken))
to verify that token.

An additional complication is that in the case of SMS, a full link to reset
passwords is not sent, but a short code. The client then asks the user to enter
this code, however the client may now not know where to send the code. Should
it send it to the identity server or the homeserver? Which sent out the code?

> In order to combat this problem, the field `submit_url` should be added in the

Reviewer 4: Can we heavily spec/imply what this URL should be please? Similar to how https://matrix.org/docs/spec/push_gateway/r0.1.0.html#post-matrix-push-v1-notify is designed - people aren't required to use that exact path, but it's pretty obvious throughout both specs that one should use that exact path.

response from both the email and msisdn variants of the `/requestToken`
Client-Server API, if and only if the verification message contains a code the
user is expected to enter into the client (for instance in the case of a short
code through SMS). If this field is omitted, the client should continue the
same behaviour from before, which is to send the token to the identity server
directly. This is intended for backwards compatibility with older servers.

If the client receives a response to `/requestToken` with `submit_url`, it

> should accept the token from user input, then make a POST request to the

Reviewer 4: ```suggestion
must accept the token from user input, then make a POST request to the
```

content of `submit_url` with the `sid`, `client_secret` and user-entered token.

> This data should be submitted as query parameters for `GET` request, and a JSON

Reviewer 4: ```suggestion
This data must be submitted as query parameters for `GET` request, and a JSON
```

I don't see where GET plays into this either.

Author: Accidentally left that in from before. Removed now.
